experimental/oak_async/tests/tests.rs

```diff
@@ -0,0 +1,287 @@
+        handles[0].set_status(ChannelReadStatus::ReadReady);
+        OakStatus::Ok
+    });
+
+    let read_result: Result<DummyData, OakError> =
+        block_on(do_channel_read(HANDLE_0)).expect("block_on failed");
+
+    assert_eq!(
+        read_result.expect("Read returned error"),
+        DummyData::new("data")
+    )
+}
+
+#[test]
+fn block_on_multiple_readers_sequentially() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        assert_eq!(handles.len(), 1);
+        let data = match handles[0].handle() {
+            HANDLE_0 => "Hello",
+            HANDLE_1 => "world!",
+            c => panic!("Unexpected waiting channel {}", c),
+        };
+        add_ready_data(handles[0].handle(), &DummyData::new(data));
+        handles[0].set_status(ChannelReadStatus::ReadReady);
+        OakStatus::Ok
+    });
+
+    let result = block_on(async {
+        let a: DummyData = do_channel_read(HANDLE_0)
+            .await
+            .expect("Failed to read channel HANDLE_0");
+        let b: DummyData = do_channel_read(HANDLE_1)
+            .await
+            .expect("Failed to read channel HANDLE_1");
+
+        [a.0, b.0].join(", ")
+    })
+    .expect("block_on failed");
+
+    assert_eq!(result, "Hello, world!");
+}
+
+#[test]
+fn block_on_multiple_readers_parallel() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        for handle in handles.iter_mut() {
+            let data = match handle.handle() {
+                HANDLE_0 => "Hello",
+                HANDLE_1 => "world!",
+                c => panic!("Unexpected waiting channel {}", c),
+            };
+            add_ready_data(handle.handle(), &DummyData::new(data));
+            handle.set_status(ChannelReadStatus::ReadReady);
+        }
+        OakStatus::Ok
+    });
+
+    let (a, b): (Result<DummyData, OakError>, Result<DummyData, OakError>) = block_on(
+        futures::future::join(do_channel_read(HANDLE_0), do_channel_read(HANDLE_1)),
+    )
+    .expect("block_on failed");
+
+    assert_eq!(
+        a.expect("Read from channel HANDLE_0 failed"),
+        DummyData::new("Hello")
+    );
+
+    assert_eq!(
+        b.expect("Read from channel HANDLE_1 failed"),
+        DummyData::new("world!")
+    );
+}
+
+#[test]
+fn block_on_drop_channel_read_immediately() {
+    init();
+    block_on(async {
+        // The future is immediately dropped, so never polled
+        let _ = do_channel_read::<DummyData>(1);
+    })
+    .expect("block_on failed");
+}
+
+#[test]
+fn block_on_drop_channel_read_after_wait() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        std::thread::sleep(std::time::Duration::from_secs(1));
+        for handle in handles.iter_mut() {
+            match handle.handle() {
+                HANDLE_0 => {
+                    handle.set_status(ChannelReadStatus::NotReady);
+                }
+                HANDLE_1 => {
+                    add_ready_data(HANDLE_1, &DummyData::new("hello"));
```

Comment:
Is it fine to modify the same state for the same handles in different test cases, which may run in parallel and in non-deterministic order?

Reply:
This is a good point. Running in parallel is fine as the state is thread local, but I did realize that if the same thread is used for multiple test cases, state could carry over from one test to the other. I'll add an `init()` function to clear the state between tests.